docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+even in the face of errors in the generated bindings. It seems to be a relatively small change, and
+by making it opt-in we avoid creating busywork for other consumers who are not urgently facing this
+problem.
+
+One downside is that consumers need to opt-in to the fix, meaning that the default behavior may still
+be surprising to new consumers. We'll mitigate this with docs and will consider revisiting the default
+behaviour if the majority of consumers adopt the new approach.
+
+This choice does also punt some potential performance improvements to future work, but that seems in keeping
+with where we are in the project's lifecycle.
+
+## Pros and Cons of the Options
+
+### Option A: Do nothing, require consumers to assume that all method calls might block.
+
+Make no changes to uniffi, and instead accept the fact that method calls are executed serially.
+Document this limitation and work with consumers to update their API definitions to account for it.
+
+* Good, because we can ship this quickly.
+* Good, because we don't give up any of the safety guarantees of the current approach.
+* Bad, because it means non-trivial rework for one of our key early consumers.
+* Bad, because it makes uniffi less attractive to potential future consumers.
+* Bad, because it keeps all the run-time overhead of the handlemap.
+* Bad, because the default behaviour still has a hidden mutex, which might be a nasty surprise
+  for future consumers in the same way that it was for Nimbus.
+
+Ultimately, we want uniffi to be a tool that helps consumers deliver value, not something that
+foist limitations upon them, which makes this option unattractive.
+
+### Option B: Let consumers mark interface definitions as threadsafe to opt in to a non-locking handlemap.
+
+Implement a variant of `ConcurrentHandleMap` that does not protect its members with a Mutex, which
+is the main source of constraints on concurrent execution in the current setup. Instead, this handlemap
+would only give out immutable references to its members, and would insist that its members are `Send`
+and `Sync` so they can be safely accessed from multiple threads.
+
+Introduce a new annotation to the UDL so that `interface` definitions can be declared as threadsafe.
+In the generated Rust scaffolding, use the new handlemap for threadsafe interfaces but keep using the
+existing `ConcurrentHandleMap` by default.
+
+* Good, because this is a relatively small change from the current behaviour, which should be fairly
+  quick to ship.
+* Good, because this is a non-breaking change for consumers who don't opt in to it.
+* Good, because the `Sync` and `Send` bounds on the underlying struct will help consumers
+  to implement their own fine-grained concurrency control while being supported by Rust's
+  compile-time guarantees.
+* Good, because consumers don't have to care about this until they discover that they need
+  fine-grained locking.
```

Comment:
"Good, because it leaves the door open to "Option C" in the future if we decide that makes sense later."?